File: PracticasPDOOJava/Practica4_Java/src/irrgarten/FuzzyPlayer.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.ArrayList;

/**
 *
 * @author Joaquín Cruz Lorenzo
 */
public class FuzzyPlayer extends Player {

    /**
     * Constructor: Crea un FuzzyPlayer a partir de otro Player (resurrección)
     * Copia su estado (nombre, inteligencia, fuerza, etc.) y sus pertenencias.
     * @param other El jugador original que ha muerto
     */
    public FuzzyPlayer(Player other) {
        super(other);
    }
    
    /**
     * Movimiento del FuzzyPlayer
     * 1) Calcula hacia donde se movería un jugador normal (preferencia)
     * 2) Pide al dado que decida si respeta esta preferencia o mueve al azar.
     */
    @Override
    public Directions move(Directions direction, Directions[] validMoves) {
        // 1. Calculo la dirección preferida usando la lógica del padre (Player)
        Directions preference = super.move(direction, validMoves);
        
        // 2. Convierto el array a ArrayList para el método Dice
        // Dice.nextStep espera un ArrayList, así que hago la conversión.
        ArrayList<Directions> validMovesList = new ArrayList<>();
        for (Directions d : validMoves) {
            validMovesList.add(d);
        }
        
        // 3. Uso el método nuevo de Dice
        return Dice.nextStep(preference, validMovesList, getIntelligence());
    }
    
    /**
     * Ataque del FuzzyPlayer
     * Suma de armas + Dice.intensity(fuerza).
     */
    @Override 
    public float attack() {
        // Nota: Necesito acceder a sumWeapons().
        return sumWeapons() + Dice.intensity(getStrength());
    }
}

File: PracticasPDOORuby/Practica4_Ruby/lib/Irrgarten/fuzzy_player.rb
# frozen_string_literal: true

require_relative 'player'
require_relative 'dice'

module Irrgarten
  class FuzzyPlayer < Player

    # Constructor a partir de otro jugador (copy)
    def initialize(other)
      # En Ruby llamamos a copy directamente, ya que initialize no tiene por qué llamar a super
      # si construimos el objeto entero nosotros.
      copy(other)
    end

    def move(direction, valid_moves)
      # 1. Calculamos la preferencia (lo que haría un jugador normal)
      # Llamamos al move del padre (Player)
      preference = super(direction, valid_moves)

      # 2. Usamos el dado para decidir
      Dice.next_step(preference, valid_moves, @intelligence)
    end

    def attack
      # Suma de armas + Dice.intensity(strength)
      sum_weapons + Dice.intensity(@strength)
    end

    def defensive_energy
      # Suma de escudos + Dice.intensity(intelligence)
      sum_shields + Dice.intensity(@intelligence)
    end

    def to_s
      "Fuzzy" + super
    end
  end
end